legion/engine/physics/physx/systems/physx_physics_system.cpp

```diff
@@ -5,56 +5,96 @@ namespace legion::physics
 {
     using namespace physx;
 
-    PxDefaultAllocator		gAllocator;
-    PxDefaultErrorCallback	gErrorCallback;
+    PhysXPhysicsSystem::physXStatics PhysXPhysicsSystem::sPhysicsStatics;
+    size_type PhysXPhysicsSystem::sSelfInstanceCounter = 0;
 
-    PxFoundation* gFoundation = nullptr;
-    PxPvd* gPvd = NULL;
-
-    PxPhysics* gPhysics = NULL;
-
-    PxDefaultCpuDispatcher* gDispatcher = NULL;
+    PxPhysics* gPhysics = nullptr;
 
     //PhysX PVD Debugger Related
     constexpr char* pvdHost = "127.0.0.1";
     constexpr size_type defaultPVDListeningPort = 5425;
     constexpr size_type defaultPVDHostTimeout = 10;
 
+    PhysXPhysicsSystem::PhysXPhysicsSystem()
+    {
+        sSelfInstanceCounter++;
+    }
+
     void PhysXPhysicsSystem::setup()
     {
-        setupPhysXGlobalVariables();
+        lazyInitPhysXVariables();
         setupDefaultScene();
-        
+
         createProcess<&PhysXPhysicsSystem::fixedUpdate>("Physics", m_timeStep);
     }
 
-    void PhysXPhysicsSystem::fixedUpdate(time::time_span<fast_time> deltaTime)
+    PhysXPhysicsSystem::~PhysXPhysicsSystem()
     {
-        m_physxScene->simulate(m_timeStep);
-        m_physxScene->fetchResults(true);
+        m_physxScene->release();
+
+        sSelfInstanceCounter--;
+
+        if (sSelfInstanceCounter == 0)
+        {
+            releasePhysXVariables();
+        }
     }
 
-    void PhysXPhysicsSystem::setupPhysXGlobalVariables()
+    void PhysXPhysicsSystem::lazyInitPhysXVariables()
     {
-        gFoundation = PxCreateFoundation(PX_PHYSICS_VERSION, gAllocator, gErrorCallback);
+        static PxDefaultAllocator sDefaultAllocator;
+        static PxDefaultErrorCallback sDefaultErrorCallback;
```

Comment:
why do we have these two if the system itself also already has these?

Reply:
mostly because I wanted all physx systems to point to the same instance of these classes, this will be changed by putting it in the new physXStatics struct